shared/src/main/scala/coop/rchain/store/LazyAdHocKeyValueCache.scala

```diff
@@ -5,42 +5,81 @@ import cats.effect.Concurrent
 import cats.effect.concurrent.{Deferred, Ref}
 import cats.syntax.all._
 
+import scala.collection.immutable.LinearSeq
+
 trait KeyValueCache[F[_], K, V] {
   def get(key: K, fallback: F[V]): F[V]
   def put(key: K, value: V): F[Unit]
   def toMap: F[Map[K, V]]
 }
 
+trait InvalidationPolicy[V] {
+  def processNewInsert(item: V): (Option[V], InvalidationPolicy[V])
+}
+
+final class NoInvalidationPolicy[V] extends InvalidationPolicy[V] {
+  override def processNewInsert(item: V): (Option[V], InvalidationPolicy[V]) = (none[V], this)
+}
+
+final case class FixedSizeLRUInvalidationPolicy[V](
+    capacity: Long,
+    useOrder: LinearSeq[V] = LinearSeq()
+) extends InvalidationPolicy[V] {
+  def processNewInsert(item: V): (Option[V], InvalidationPolicy[V]) =
+    useOrder match {
+      case Seq() => (none[V], FixedSizeLRUInvalidationPolicy(capacity, useOrder))
+      case v     => (v.head.some, FixedSizeLRUInvalidationPolicy(capacity, useOrder.tail :+ item))
```

Comment:
Append operation on a Seq is linear so not the best choice here.
https://docs.scala-lang.org/overviews/collections-2.13/performance-characteristics.html

Reply:
I see, I used linear seq because https://docs.scala-lang.org/overviews/collections/seqs.html says it's optimised for `head` and `tail`.

Reply:
That's true, but you are also using `:+` (append) which will iterate the whole list.